Sources/SwiftParser/TopLevel.swift

```diff
@@ -98,7 +98,7 @@ extension Parser {
   /// This function is used when parsing places where function bodies are
   /// optional - like the function requirements in protocol declarations.
   mutating func parseOptionalCodeBlock(allowInitDecl: Bool = true) -> RawCodeBlockSyntax? {
-    guard self.at(.leftBrace) else {
+    guard let _ = self.canRecoverTo(TokenSpec(.leftBrace)) else {
```

Comment:
I can't give you an exact counter-example but I'm a little hesitant to allow recovering to a left brace on an arbitrary line because that could cause us to skip over arbitrary identifiers to reach a `{` that is really completely unrelated to the optional code block we want to parse. E.g. if you have

```swift
var x: Int
somethingThatTakesATrailingClosure {}
```

we don't want to consume `somethingThatTakesATrailingClosure` as unexpected to recover to the `{`. I haven't checked this exact example but I hope the idea comes across. I would feel a lot more comfortable if we only allowed recovery to `{` on the same line because that way the `{` can't belong to a different statement since statements on the same line need to be separated by `;` and `;` has a higher precedence than `{`.

I.e. I'd suggest to change this to something like the following

```suggestion
    guard self.at(.leftBrace) || self.canRecoverTo(TokenSpec(.leftBrace, allowAtStartOfLine: false)) != nil else {
```

same above when looking for `->`